File: endpoints/csharp/ExampleSync.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ZeroDDS Contributors
//
// Deep example (sync): a realistic sensor-telemetry flow. A publisher frames
// five typed Reading(Id, Value, Label) samples and delivers them; the
// subscriber owns the run-loop and polls, decoding EVERY field byte-for-byte.

using System;
using ZeroDDS;

namespace ZeroDDS.Examples
{
    public static class ExampleSync
    {
        // Reading { uint32 Id; float Value; string Label }.
        private readonly struct Reading
        {
            public readonly uint Id;
            public readonly float Value;
            public readonly string Label;

            public Reading(uint id, float value, string label)
            {
                Id = id;
                Value = value;
                Label = label;
            }

            public byte[] Marshal(Endianness endian)
            {
                var w = new Writer(endian);
                w.PutU32(Id);
                w.PutF32(Value);
                w.PutString(Label);
                return w.Bytes();
            }

            public static Reading Decode(byte[] body)
            {
                var r = new Reader(body, Endianness.Little);
                return new Reading(r.GetU32(), r.GetF32(), r.GetString());
            }
        }

        public static int Main()
        {
            const int total = 5;
            var transport = new MemTransport();
            var client = new Client(transport);

            for (int i = 0; i < total; i++)
            {
                var rd = new Reading((uint)(0x1000 + i), 20.0f + i * 0.5f, $"bay-{i:D2}");
                client.Write(rd.Marshal(Endianness.Little));
            }

            int got = 0;
            while (got < total)
            {
                var body = client.Poll();
                if (body == null) break;
                var rd = Reading.Decode(body);
                Console.WriteLine($"sync reading {got}: id=0x{rd.Id:x} value={rd.Value:F1} label=\"{rd.Label}\"");
                got++;
            }

            if (got != total)
            {
                Console.Error.WriteLine($"incomplete: got {got} of {total}");
                return 1;
            }
            Console.WriteLine("ALL OK");
            return 0;
        }
    }
}

File: endpoints/python/example_sync.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 ZeroDDS Contributors
#
# Deep example (sync): a realistic sensor-telemetry flow. A publisher frames
# five typed Reading(id, value, label) samples and delivers them; the subscriber
# owns the run-loop and polls, decoding EVERY field byte-for-byte.

import sys

import zerodds_wire as zw
import zerodds_endpoint as ze


class Reading(object):
    __slots__ = ("id", "value", "label")

    def __init__(self, id, value, label):
        self.id = id
        self.value = value
        self.label = label

    def marshal(self, endian):
        w = zw.Writer(endian)
        w.put_u32(self.id)
        w.put_f32(self.value)
        w.put_string(self.label)
        return w.bytes()

    @staticmethod
    def decode(body):
        r = zw.Reader(body, zw.LE)
        return Reading(r.get_u32(), r.get_f32(), r.get_string())


def main():
    total = 5
    transport = ze.MemTransport()
    client = ze.Client(transport)
    for i in range(total):
        r = Reading(0x1000 + i, 20.0 + i * 0.5, "bay-%02d" % i)
        client.write(r.marshal(zw.LE))

    got = 0
    while got < total:
        body = client.poll()
        if body is None:
            break
        r = Reading.decode(body)
        print('sync reading %d: id=0x%x value=%.1f label="%s"'
              % (got, r.id, r.value, r.label))
        got += 1

    if got != total:
        sys.stderr.write("incomplete: got %d of %d\n" % (got, total))
        return 1
    print("ALL OK")
    return 0


if __name__ == "__main__":
    sys.exit(main())
